micall/core/aln2counts.py

```diff
@@ -123,25 +123,171 @@ def trim_contig_name(contig_name):
                     logger.debug("Continuing with dict counts.")
-    return nuc_dict
+
+
+def combine_region_insertions(insertions_dict, region_insertions, region_start):
+    assert region_start is not None
+    if region_insertions is None:
+        return insertions_dict
+    new_insertions = {}
+    for position in region_insertions.keys():
+        ref_position = position + region_start - 1
+        if ref_position not in insertions_dict:
+            shifted_positions = (ref_position-3, ref_position-2, ref_position-1, ref_position+1, ref_position+2,
+                                 ref_position+3)
+            if any(shift_pos in insertions_dict for shift_pos in shifted_positions):
+                logger.debug(f"Disagreement or shift in insertions between regions. Position {ref_position + 1}")
+            else:
+                new_insertions[ref_position] = region_insertions[position]
+        else:
+            if insertions_dict[ref_position].items() != region_insertions[position].items():
+                logger.debug(f"Insertion counts don't match up. Position {ref_position + 1}")
+                logger.debug("Continuing with dict counts.")
+    insertions_dict.update(new_insertions)
+
+
+def insert_insertions(insertions, consensus_nucs):
+    for position in insertions.keys():
+        neighbour_left = consensus_nucs.get(position)
+        neighbour_right = consensus_nucs.get(position + 1)
+        coverage_left = neighbour_left.get_coverage() if neighbour_left is not None else 0
+        coverage_right = neighbour_right.get_coverage() if neighbour_left is not None else 0
+        coverage_nuc = max(coverage_left, coverage_right)
+        num_insertions = len(insertions[position])
+        for i in range(num_insertions):
+            # use non-integer positions for insertion nucleotides
+            insertion_position = position + (i + 1) / (num_insertions + 1)
+            coverage_insertion = insertions[position][i].get_coverage()
+            coverage_no_insertion = coverage_nuc - coverage_insertion
+            if coverage_no_insertion > 0:
+                insertions[position][i].counts['-'] = coverage_no_insertion
+            consensus_nucs[insertion_position] = insertions[position][i]
+    return consensus_nucs
+
+
+def aggregate_insertions(insertions_counter, coverage_nuc=0, consensus_pos=None):
+    aggregated_insertions = defaultdict()
+
+    if len(insertions_counter) == 0:
+        return aggregated_insertions
+
+    length = 0
+    for insertion in insertions_counter:
+        if len(insertion) > length:
+            length = len(insertion)
+
+    for i in range(length):
+        insertion_nuc = SeedNucleotide()
+        insertion_nuc.consensus_index = consensus_pos
+        for insertion, count in insertions_counter.items():
+            if len(insertion) > i:
+                insertion_nuc.count_nucleotides(insertion[i], count)
+        coverage_insertion = insertion_nuc.get_coverage(count_nucs='ACGT')
```

Comment:
In case there are already deletion counts in the insertions, we would want to count them towards the non-insertion coverage and not the insertion counts. I'm not quite sure whether this could be the case in weird border cases at this point - but I thought of a simpler implementation that works in any case and where we don't have to restrict the nucleotides!